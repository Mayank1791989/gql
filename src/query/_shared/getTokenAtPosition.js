/* @flow */
import _getTokenAtPosition from '../../shared/getTokenAtPosition';
// import RelayQLParser from './shared/Parser';
import type { ParserConfig } from '../_shared/types';
import type { Position, IParser } from '../../shared/types';

function newParser(parserConfig: ParserConfig): IParser {
  const _parserConfig = typeof parserConfig === 'string' ? [parserConfig] : parserConfig;
  const [name, opts] = _parserConfig;
  // $FlowDisableNextLine
  const Parser = require(`./Parsers/${name}`).default;
  return new Parser(opts);
}

export const getTokenAtPosition = (
  sourceText: string,
  position: Position,
  parserConfig: ParserConfig,
) => {
  const parser = newParser(parserConfig);
  const token = _getTokenAtPosition(parser, sourceText, position);
  return token;
};
export default getTokenAtPosition;
