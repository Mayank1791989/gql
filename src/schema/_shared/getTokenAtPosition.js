/* @flow */
import SchemaParser from './SchemaParser';
import _getTokenAtPosition from '../../shared/getTokenAtPosition';
import type { Position } from '../../shared/types';

export const getTokenAtPosition = (sourceText: string, position: Position) => {
  const parser = new SchemaParser();
  return _getTokenAtPosition(parser, sourceText, position);
};
export default getTokenAtPosition;
