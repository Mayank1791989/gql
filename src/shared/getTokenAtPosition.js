/* @flow */
/* eslint-disable no-loop-func */
import type { Position, Token, IParser } from './types';
import CharacterStream from 'codemirror-graphql/utils/CharacterStream';
import invariant from 'graphql/jsutils/invariant';
import printTokenState from './printTokenState';
import splitLines from './splitLines';
import whileSafe from './whileSafe';

export default function getTokenAtPosition(
  parser: IParser,
  sourceText: string,
  position: Position,
): Token {
  const state = parser.startState();
  const lines = splitLines(sourceText);
  const { line, column } = position;
  const lastLineIndex = Math.min(line, lines.length) - 1;

  let style;
  let stream;

  for (let index = 0; index <= lastLineIndex; index += 1) {
    stream = new CharacterStream(lines[index]);
    if (index < lastLineIndex) { // if not last line
      whileSafe({
        condition: () => !stream.eol(),
        call: () => {
          style = parser.token(stream, state);
          // console.log(
          //   `line: ${index} [${lines[index]}]`,
          //   `style: [${style}] token: [${stream.current()}] \n`,
          //   'state: ', state, '\n', printTokenState(state),
          // );
        },
        logOnInfiniteLoop: () => {
          console.log(
            `line: ${index} [${lines[index]}]`,
            `style: [${style}] token: [${stream.current()}] \n`,
            'state: ', state, '\n', printTokenState(state),
          );
        },
      });
    } else if (index === lastLineIndex) { // if last line run till column
      whileSafe({
        condition: () => stream.getCurrentPosition() < column && !stream.eol(),
        call: () => {
          style = parser.token(stream, state);
          // console.log(
          //   `line: ${index} [${lines[index]}]`,
          //   `style: [${style}] token: [${stream.current()}] \n`,
          //   'state: ', state, '\n', printTokenState(state),
          // );
        },
        logOnInfiniteLoop: () => {
          console.log(
            `line: ${index} [${lines[index]}]`,
            `style: [${style}] token: [${stream.current()}] \n`,
            'state: ', state, '\n', printTokenState(state),
          );
        },
      });
    }
  }

  invariant(style, 'expected style should have some value');
  invariant(stream, 'expected stream should have some value');

  const pos = Math.min(stream.getCurrentPosition(), column - 1);

  return {
    start: stream.getStartOfToken(),
    end: stream.getCurrentPosition(),
    string: stream.current(),
    prevChar: stream._sourceText.charAt(pos - 1),
    state,
    style,
  };
}
