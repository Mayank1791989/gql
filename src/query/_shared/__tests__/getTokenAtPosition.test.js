/* @flow */
import getTokenAtPosition from '../getTokenAtPosition';

const relayQLParser = ['EmbeddedQueryParser', { startTag: 'Relay.QL`', endTag: '`' }];

test('should able to parse Relay.QL tags', () => {
  const source = `
    const a = Relay.QL\`
      fragment on Test {
        name
        image
      }
    \`;
  `;

  const token = getTokenAtPosition(source, { line: 4, column: 21 }, relayQLParser);
  expect(token).toMatchSnapshot();
});

test('should able to parse multiple Relay.QL tags in single file', () => {
  const source = `
    const a = Relay.QL\`
      fragment on Test { image }
    \`;

    const b = Relay.QL\`
      fragment on Test { image }
    \`;
  `;

  const token = getTokenAtPosition(source, { line: 7, column: 21 }, relayQLParser);
  expect(token).toMatchSnapshot();
});

describe('should able to parse Relay.QL queries with interpolation', () => {
  it('single line', () => {
    const source = `
      const a = Relay.QL\`
        fragment Name on Test {
          \${abcd}
          name
        }
      \`;
    `;

    const token = getTokenAtPosition(source, { line: 5, column: 12 }, relayQLParser);
    expect(token).toMatchSnapshot();
  });

  it('multi line', () => {
    const source = `
      const a = Relay.QL\`
        fragment Name on Test {
          \${
              abcd
           }
           name
        }
      \`;
    `;

    const token = getTokenAtPosition(source, { line: 7, column: 14 }, relayQLParser);
    expect(token).toMatchSnapshot();
  });

  it('with curly braces inside interpolation', () => {
    const source = `
      const a = Relay.QL\`
        fragment Name on Test {
          \${abcd.getFramgent({
             var1: variables.var1
          })}
          name
        }
      \`;
    `;

    const token = getTokenAtPosition(source, { line: 7, column: 14 }, relayQLParser);
    expect(token).toMatchSnapshot();
  });

  it('[Bug] should able to parse fields after consecutive interpolated fragments', () => {
    const source = `
      const a = Relay.QL\`
        fragment XYZ on Test {
          \${fragment}
          \${fragment}
          name
        }
      \`;
    `;

    expect(() => getTokenAtPosition(source, { line: 6, column: 14 }, relayQLParser)).not.toThrow();
  });
});
