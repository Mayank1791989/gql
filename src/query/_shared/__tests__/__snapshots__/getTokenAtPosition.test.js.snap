exports[`should able to parse Relay.QL queries with interpolation multi line 1`] = `
Object {
  "end": 15,
  "prevChar": "a",
  "start": 11,
  "state": Object {
    "indentLevel": 5,
    "jsInlineFragment": null,
    "kind": "Field",
    "level": 0,
    "levels": Array [
      5,
    ],
    "name": "name",
    "needsAdvance": true,
    "needsSeperator": false,
    "prevState": Object {
      "indentLevel": 5,
      "jsInlineFragment": null,
      "kind": "Selection",
      "level": 0,
      "levels": Array [
        5,
      ],
      "name": null,
      "needsAdvance": false,
      "needsSeperator": false,
      "prevState": Object {
        "indentLevel": 5,
        "jsInlineFragment": null,
        "kind": "SelectionSet",
        "level": 0,
        "levels": Array [
          5,
        ],
        "name": null,
        "needsAdvance": false,
        "needsSeperator": false,
        "prevState": Object {
          "indentLevel": 4,
          "kind": "FragmentDefinition",
          "level": 0,
          "levels": Array [
            5,
          ],
          "name": "Name",
          "needsAdvance": false,
          "needsSeperator": false,
          "prevState": Object {
            "indentLevel": 4,
            "kind": "Definition",
            "level": 0,
            "name": null,
            "needsSeperator": false,
            "prevState": Object {
              "indentLevel": 4,
              "kind": "Document",
              "level": 0,
              "name": null,
              "needsSeperator": false,
              "prevState": Object {
                "level": 0,
              },
              "rule": Array [
                "Definition",
              ],
              "step": 0,
              "type": null,
            },
            "rule": [Function],
            "step": 0,
            "type": null,
          },
          "rule": Array [
            Object {
              "match": [Function],
              "style": "keyword",
            },
            Object {
              "ofRule": Object {
                "match": [Function],
                "style": "def",
                "update": [Function],
              },
            },
            "TypeCondition",
            Object {
              "isList": true,
              "ofRule": "Directive",
              "separator": undefined,
            },
            "SelectionSet",
          ],
          "step": 4,
          "type": "Test",
        },
        "rule": Array [
          Object {
            "match": [Function],
            "style": "punctuation",
          },
          Object {
            "isList": true,
            "ofRule": "Selection",
            "separator": undefined,
          },
          Object {
            "match": [Function],
            "style": "punctuation",
          },
        ],
        "step": 1,
        "type": null,
      },
      "rule": [Function],
      "step": 0,
      "type": null,
    },
    "rule": Array [
      Object {
        "match": [Function],
        "style": "property",
        "update": [Function],
      },
      Object {
        "ofRule": "Arguments",
      },
      Object {
        "isList": true,
        "ofRule": "Directive",
        "separator": undefined,
      },
      Object {
        "ofRule": "SelectionSet",
      },
    ],
    "step": 0,
    "type": null,
  },
  "string": "name",
  "style": "property",
}
`;

exports[`should able to parse Relay.QL queries with interpolation single line 1`] = `
Object {
  "end": 14,
  "prevChar": "n",
  "start": 10,
  "state": Object {
    "indentLevel": 5,
    "jsInlineFragment": null,
    "kind": "Field",
    "level": 0,
    "levels": Array [
      5,
    ],
    "name": "name",
    "needsAdvance": true,
    "needsSeperator": false,
    "prevState": Object {
      "indentLevel": 5,
      "jsInlineFragment": null,
      "kind": "Selection",
      "level": 0,
      "levels": Array [
        5,
      ],
      "name": null,
      "needsAdvance": false,
      "needsSeperator": false,
      "prevState": Object {
        "indentLevel": 5,
        "jsInlineFragment": null,
        "kind": "SelectionSet",
        "level": 0,
        "levels": Array [
          5,
        ],
        "name": null,
        "needsAdvance": false,
        "needsSeperator": false,
        "prevState": Object {
          "indentLevel": 4,
          "kind": "FragmentDefinition",
          "level": 0,
          "levels": Array [
            5,
          ],
          "name": "Name",
          "needsAdvance": false,
          "needsSeperator": false,
          "prevState": Object {
            "indentLevel": 4,
            "kind": "Definition",
            "level": 0,
            "name": null,
            "needsSeperator": false,
            "prevState": Object {
              "indentLevel": 4,
              "kind": "Document",
              "level": 0,
              "name": null,
              "needsSeperator": false,
              "prevState": Object {
                "level": 0,
              },
              "rule": Array [
                "Definition",
              ],
              "step": 0,
              "type": null,
            },
            "rule": [Function],
            "step": 0,
            "type": null,
          },
          "rule": Array [
            Object {
              "match": [Function],
              "style": "keyword",
            },
            Object {
              "ofRule": Object {
                "match": [Function],
                "style": "def",
                "update": [Function],
              },
            },
            "TypeCondition",
            Object {
              "isList": true,
              "ofRule": "Directive",
              "separator": undefined,
            },
            "SelectionSet",
          ],
          "step": 4,
          "type": "Test",
        },
        "rule": Array [
          Object {
            "match": [Function],
            "style": "punctuation",
          },
          Object {
            "isList": true,
            "ofRule": "Selection",
            "separator": undefined,
          },
          Object {
            "match": [Function],
            "style": "punctuation",
          },
        ],
        "step": 1,
        "type": null,
      },
      "rule": [Function],
      "step": 0,
      "type": null,
    },
    "rule": Array [
      Object {
        "match": [Function],
        "style": "property",
        "update": [Function],
      },
      Object {
        "ofRule": "Arguments",
      },
      Object {
        "isList": true,
        "ofRule": "Directive",
        "separator": undefined,
      },
      Object {
        "ofRule": "SelectionSet",
      },
    ],
    "step": 0,
    "type": null,
  },
  "string": "name",
  "style": "property",
}
`;

exports[`should able to parse Relay.QL queries with interpolation with curly braces inside interpolation 1`] = `
Object {
  "end": 14,
  "prevChar": "m",
  "start": 10,
  "state": Object {
    "indentLevel": 5,
    "jsInlineFragment": null,
    "kind": "Field",
    "level": 0,
    "levels": Array [
      5,
    ],
    "name": "name",
    "needsAdvance": true,
    "needsSeperator": false,
    "prevState": Object {
      "indentLevel": 5,
      "jsInlineFragment": null,
      "kind": "Selection",
      "level": 0,
      "levels": Array [
        5,
      ],
      "name": null,
      "needsAdvance": false,
      "needsSeperator": false,
      "prevState": Object {
        "indentLevel": 5,
        "jsInlineFragment": null,
        "kind": "SelectionSet",
        "level": 0,
        "levels": Array [
          5,
        ],
        "name": null,
        "needsAdvance": false,
        "needsSeperator": false,
        "prevState": Object {
          "indentLevel": 4,
          "kind": "FragmentDefinition",
          "level": 0,
          "levels": Array [
            5,
          ],
          "name": "Name",
          "needsAdvance": false,
          "needsSeperator": false,
          "prevState": Object {
            "indentLevel": 4,
            "kind": "Definition",
            "level": 0,
            "name": null,
            "needsSeperator": false,
            "prevState": Object {
              "indentLevel": 4,
              "kind": "Document",
              "level": 0,
              "name": null,
              "needsSeperator": false,
              "prevState": Object {
                "level": 0,
              },
              "rule": Array [
                "Definition",
              ],
              "step": 0,
              "type": null,
            },
            "rule": [Function],
            "step": 0,
            "type": null,
          },
          "rule": Array [
            Object {
              "match": [Function],
              "style": "keyword",
            },
            Object {
              "ofRule": Object {
                "match": [Function],
                "style": "def",
                "update": [Function],
              },
            },
            "TypeCondition",
            Object {
              "isList": true,
              "ofRule": "Directive",
              "separator": undefined,
            },
            "SelectionSet",
          ],
          "step": 4,
          "type": "Test",
        },
        "rule": Array [
          Object {
            "match": [Function],
            "style": "punctuation",
          },
          Object {
            "isList": true,
            "ofRule": "Selection",
            "separator": undefined,
          },
          Object {
            "match": [Function],
            "style": "punctuation",
          },
        ],
        "step": 1,
        "type": null,
      },
      "rule": [Function],
      "step": 0,
      "type": null,
    },
    "rule": Array [
      Object {
        "match": [Function],
        "style": "property",
        "update": [Function],
      },
      Object {
        "ofRule": "Arguments",
      },
      Object {
        "isList": true,
        "ofRule": "Directive",
        "separator": undefined,
      },
      Object {
        "ofRule": "SelectionSet",
      },
    ],
    "step": 0,
    "type": null,
  },
  "string": "name",
  "style": "property",
}
`;

exports[`test should able to parse Relay.QL tags 1`] = `
Object {
  "end": 12,
  "prevChar": "e",
  "start": 8,
  "state": Object {
    "indentLevel": 4,
    "kind": "Field",
    "level": 0,
    "levels": Array [
      4,
    ],
    "name": "name",
    "needsAdvance": true,
    "needsSeperator": false,
    "prevState": Object {
      "indentLevel": 4,
      "kind": "Selection",
      "level": 0,
      "levels": Array [
        4,
      ],
      "name": null,
      "needsAdvance": false,
      "needsSeperator": false,
      "prevState": Object {
        "indentLevel": 4,
        "kind": "SelectionSet",
        "level": 0,
        "levels": Array [
          4,
        ],
        "name": null,
        "needsAdvance": false,
        "needsSeperator": false,
        "prevState": Object {
          "indentLevel": 3,
          "kind": "FragmentDefinition",
          "level": 0,
          "levels": Array [
            4,
          ],
          "name": null,
          "needsAdvance": false,
          "needsSeperator": false,
          "prevState": Object {
            "indentLevel": 3,
            "kind": "Definition",
            "level": 0,
            "name": null,
            "needsSeperator": false,
            "prevState": Object {
              "indentLevel": 3,
              "kind": "Document",
              "level": 0,
              "name": null,
              "needsSeperator": false,
              "prevState": Object {
                "level": 0,
              },
              "rule": Array [
                "Definition",
              ],
              "step": 0,
              "type": null,
            },
            "rule": [Function],
            "step": 0,
            "type": null,
          },
          "rule": Array [
            Object {
              "match": [Function],
              "style": "keyword",
            },
            Object {
              "ofRule": Object {
                "match": [Function],
                "style": "def",
                "update": [Function],
              },
            },
            "TypeCondition",
            Object {
              "isList": true,
              "ofRule": "Directive",
              "separator": undefined,
            },
            "SelectionSet",
          ],
          "step": 4,
          "type": "Test",
        },
        "rule": Array [
          Object {
            "match": [Function],
            "style": "punctuation",
          },
          Object {
            "isList": true,
            "ofRule": "Selection",
            "separator": undefined,
          },
          Object {
            "match": [Function],
            "style": "punctuation",
          },
        ],
        "step": 1,
        "type": null,
      },
      "rule": [Function],
      "step": 0,
      "type": null,
    },
    "rule": Array [
      Object {
        "match": [Function],
        "style": "property",
        "update": [Function],
      },
      Object {
        "ofRule": "Arguments",
      },
      Object {
        "isList": true,
        "ofRule": "Directive",
        "separator": undefined,
      },
      Object {
        "ofRule": "SelectionSet",
      },
    ],
    "step": 0,
    "type": null,
  },
  "string": "name",
  "style": "property",
}
`;

exports[`test should able to parse multiple Relay.QL tags in single file 1`] = `
Object {
  "end": 22,
  "prevChar": "e",
  "start": 18,
  "state": Object {
    "indentLevel": 3,
    "kind": "TypeCondition",
    "level": 0,
    "name": null,
    "needsAdvance": true,
    "needsSeperator": false,
    "prevState": Object {
      "indentLevel": 3,
      "kind": "FragmentDefinition",
      "level": 0,
      "name": null,
      "needsAdvance": false,
      "needsSeperator": false,
      "prevState": Object {
        "indentLevel": 3,
        "kind": "Definition",
        "level": 0,
        "name": null,
        "needsSeperator": false,
        "prevState": Object {
          "indentLevel": 3,
          "kind": "Document",
          "level": 0,
          "name": null,
          "needsSeperator": false,
          "prevState": Object {
            "level": 0,
          },
          "rule": Array [
            "Definition",
          ],
          "step": 0,
          "type": null,
        },
        "rule": [Function],
        "step": 0,
        "type": null,
      },
      "rule": Array [
        Object {
          "match": [Function],
          "style": "keyword",
        },
        Object {
          "ofRule": Object {
            "match": [Function],
            "style": "def",
            "update": [Function],
          },
        },
        "TypeCondition",
        Object {
          "isList": true,
          "ofRule": "Directive",
          "separator": undefined,
        },
        "SelectionSet",
      ],
      "step": 2,
      "type": "Test",
    },
    "rule": Array [
      Object {
        "match": [Function],
        "style": "keyword",
      },
      Object {
        "match": [Function],
        "style": "atom",
        "update": [Function],
      },
    ],
    "step": 1,
    "type": null,
  },
  "string": "Test",
  "style": "atom",
}
`;
